frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1139,7 +1139,10 @@ func dispatchCopy(d *dispatchState, cfg copyConfig) error {
 	}
 
 	if cfg.params.DestPath == "." || cfg.params.DestPath == "" || cfg.params.DestPath[len(cfg.params.DestPath)-1] == filepath.Separator {
```

Comment:
Would it be bit more correct if this condition also checked:

```
lastChar := cfg.params.DestPath[len(cfg.params.DestPath)-1]
isWindows :=
if cfg.params.DestPath == "." || cfg.params.DestPath == "" || lastChar == "/" || (lastChar == "\" && isWindows) {
```

Reply:
BTW, given what is doing at `L1136` , do we actually need this block now or we just continue with `dest` from here onwards?

https://github.com/moby/buildkit/blob/fb97e6defdc171d943c304565eeb72794ea0a648/frontend/dockerfile/dockerfile2llb/convert.go#L1550-L1568

Reply:
Afaics, `pathRelativeToWorkingDir` calls `system.NormalizePath()` and `NormalizePath()` takes `keepSlash bool`, but `pathRelativeToWorkingDir` is setting it to `false`. If we would set it to `true` instead then it looks like we don't need this extra block.

cc @gabriel-samfira to double-check the intention of `keepSlash=false`

Reply:
Perhaps this was for some old code, since? there's no where `system.NormalizePath()` is called with `keepSlash=true`, as far as I can see. At some point, may be we should do away with that param?
```
$ grep -rn "\.NormalizePath("
frontend/dockerfile/dockerfile2llb/convert.go:1279:                             pattern, err = system.NormalizePath("/", pattern, d.platform.OS, false)
frontend/dockerfile/dockerfile2llb/convert.go:1287:                     src, err = system.NormalizePath("/", src, d.platform.OS, false)
frontend/dockerfile/dockerfile2llb/convert.go:1565:             return system.NormalizePath("/", p, platform.OS, false)
frontend/dockerfile/dockerfile2llb/convert.go:1567:     return system.NormalizePath(dir, p, platform.OS, false)
```

Reply:
I think I would rather keep `keepSlash` or add a new helper function for this functionality (if there is only one caller then it can remain private). The current long if statement is not very robust, can't be tested and is easy to break.

Reply:
the `keepSlash` logic was carried forward from the function that `NormalizePath()` replaced.

Looking over the rest of the PR now.

Reply:
I think it's safe to switch `NormalizePath` to `true` in `pathRelativeToWorkingDir`. That function is only used once, in `convert.go`. And we seem to be duplicating some of its logic in that if statement that follows right after we call `pathRelativeToWorkingDir()`. Let's just remove that if block and flip that boolean in `pathRelativeToWorkingDir`

Reply:
Thank @gabriel-samfira for the review. I've made the change PTAL.